### PR TITLE
Removed the technical debt by upgrading the versions of almost all of the out of date dependencies 

### PR DESCRIPTION
**Pull request by andyglick**
Guy and company. I noticed that the open source code base had accumulated a lot of technical debt as the dependencies for many 3rd party jars weren't up to date. I felt that the easiest way that I could contribute to the code base was to remove as much of the debt as seemed feasible. I updated almost every dependency that had an existing upgrade which was compatible with its module. That means that the hibernate3 module still references hibernate 3 and the hibernate4 module still references hibernate 4. 

You might want to consider adding a hibernate 5 module as hibernate 4 is now a legacy release.

The only exception to the universal upgrading was org.apache.logging.log4j:log4j-core in the util module. When I bumped its version from 2.2 to 2.8 tests broke and since you all are doing things in the util module with loggers that I don't really understand, I decided to leave well enough alone.  

I added a number of plugins to the build, the maven-dependency-plugin, the maven-enforcer-plugin, the maven-help-plugin and the versions-maven-plugin. I use these as parts of my toolset when I work at reducing the technical debt often found in Maven projects, my own and others.

Github reports that my code can be merged onto your existing master branch without conflicts. All of the tests pass. There are some issues with Javadoc generation, it looks as if some of your comment conventions are causing the Javadoc exercises to produce warnings or even crash, but these issues don't keep the build from completing.

Please let me know if you have any issues or concerns with my efforts.

Thank you all for contributing this really fine effort to the open source community.  

BTW my email address is andyglick@gmail.com if you have to send me a contributor form.

### CODE REVIEW

**Comment by GuyPardon** [comment]
Andy,

Received with thanks!

Let us have a look and get back to you if we have questions.

Guy

**Comment by pascalleclercq** [comment]
Hi @andyglick,
thank you for your contributions,

we need to double-check It  because some are interesting and points out things we forgot to update/upgrade or even remove from the pom.xml.

On the other hand, we are reluctant of upgrading third parties dependencies to the latests available because It's only compilation dependency for us and we don't want to take the risk to depend on the latests features that would force our users to upgrade on this peculiar dependency (e.g. slf4j).

AFAIK most third party dependencies we use are pretty good at upward compatibility so It's safe to keep them pretty outdated.

**Comment by andyglick** [comment]
Guy, thank you for the question abut the jboss-transactions dependency. I wasn't paying enough attention when I chose it. As you can see I actually prefer the latest javax.transaction library.

Pascall, I don't offer many "finished products" for use by others, and for the 1 that I do maintain, knife-api, I always keep its primary dependencies up to date. But I do understand your concerns.

With regard to the Java changes that I made, they were called out by the errorprone Java compiler from Google. You might want to take a look at how I have configured it in the root pom. I set it up as a profile. You might want to do something similar, it finds and flags issues that javac ignores.

File: public/util/src/test/java/com/atomikos/logging/LoggerFactoryTestJUnit.java [hunk 1]
```diff
@@ -8,14 +8,20 @@
...
+
 import junit.framework.TestCase;
 
-public class LoggerFactoryTestJUnit extends TestCase {
+import static org.junit.Assert.assertEquals;
+
+public class LoggerFactoryTestJUnit {
 
+  @Test
+  @Ignore
```
Comment (pascalleclercq) [inline public/util/src/test/java/com/atomikos/logging/LoggerFactoryTestJUnit.java]: Why @Ignore here ?

File: public/transactions/src/main/java/com/atomikos/finitestates/Stateful.java [hunk 2]
```diff
@@ -10,11 +10,13 @@
 
 import com.atomikos.recovery.TxState;
 
+import java.io.Serializable;
 
-public interface Stateful{
+public interface Stateful extends Serializable
```
Comment (pascalleclercq) [inline public/transactions/src/main/java/com/atomikos/finitestates/Stateful.java]: Why "Stateful" should be "Serializable" ?

File: public/transactions/src/test/java/com/atomikos/icatch/config/ConfigurationTestJUnit.java [hunk 3]
```diff
@@ -77,6 +78,7 @@ public void testSystemPropertyOverridesJtaPropertiesFile() {
 	}
 	
 	@Test
+  @Ignore
```
Comment (pascalleclercq) [inline public/transactions/src/test/java/com/atomikos/icatch/config/ConfigurationTestJUnit.java]: Why @Ignore here ?

File: public/util/src/main/java/com/atomikos/timing/AlarmTimer.java [hunk 4]
```diff
@@ -5,16 +5,16 @@
...
 
+import java.io.Serializable;
 
 /**
  * A common interface for timers.
  * 
  * @author Lars J. Nilsson
  */
-public interface AlarmTimer extends Runnable {
+public interface AlarmTimer extends Serializable, Runnable {
```
Comment (pascalleclercq) [inline public/util/src/main/java/com/atomikos/timing/AlarmTimer.java]: Why Serializable here ?

**Comment by GuyPardon** [comment]
Andy,

If you have some time, please reply to Pascal's code review questions so we can progress on this pull request - thanks!

**Pull request closed without merging**